src/charm.py

```diff
@@ -844,42 +707,31 @@ def _set_pebble_plan(self) -> None:
             self._container.replan()
             logger.info("Pebble layer added")
 
-    def _restore_vault(self, snapshot: StreamingBody) -> bool:
-        """Restore vault using a raft snapshot.
-
-        Args:
-            snapshot: Snapshot to be restored as a StreamingBody from the S3 storage.
+    def _get_active_vault_client(self) -> VaultClient | None:
+        """Return a client for the _active_ vault service.
 
-        Returns:
-            bool: True if the restore was successful, False otherwise.
+        This may not be the Vault service running on this unit.
         """
```

Comment:
I wonder if there is a way to synchronize the juju leader with the vault leader.

If there was, we wouldn't need this logic since we could only allow this action to run on the leader (and hence, avoid the transmission of the snapshot to another unit). Or, should we force them to run this on the unit that is the Vault leader?

Reply:
I do not think there is a way to synchronize the juju leader with the vault leader, no.

I believe this is the simplest approach for users. I would keep it that way, even if it is less efficient.